consensus/currency.go

```diff
@@ -0,0 +1,114 @@
+package consensus
+
+import (
+	"errors"
+	"math/big"
+)
+
+var (
+	ErrOverflow = errors.New("Currency overflowed 128 bits")
+)
+
+// A Currency is a 128-bit unsigned integer. Currency operations are performed
+// via math/big.
+//
+// The Currency object also keeps track of whether an overflow has occurred
+// during arithmetic operations. Once the 'overflow' flag has been set to
+// true, all subsequent operations will return an error, and the result of the
+// operation is undefined. This flag can never be reset; a new Currency must
+// be created. Callers can also manually check for overflow using the Overflow
+// method.
+type Currency struct {
+	i  big.Int
+	of bool // has an overflow ever occurred?
+}
+
+func NewCurrency(x uint64) Currency {
+	// no possibility of error
+	c, _ := BigToCurrency(new(big.Int).SetUint64(x))
+	return c
+}
+
+func BigToCurrency(b *big.Int) (c Currency, err error) {
+	if b.BitLen() > 128 || b.Sign() < 0 {
+		c.of = true
+		err = ErrOverflow
+		return
+	}
+	c.i = *b
+	return
+}
+
+func (c *Currency) SetBig(b *big.Int) (err error) {
+	oldOF := c.of
+	*c, err = BigToCurrency(b)
+	c.of = c.of || oldOF // preserve overflow flag
+	return
+}
```

Comment:
On a related note, I thought about just unexporting the `Currency` type entirely. It used to use `*big.Int` instead of a `big.Int`, and it caused nil dereference panics because I was writing `var sum Currency` instead of `sum := NewCurrency(0)`.
It should be safe to write it either way now, but unexporting would give us total control over how `Currency` objects are created and manipulated.